airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,354 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+from urllib.parse import parse_qsl, urlparse
+
+import pendulum
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import Oauth2Authenticator
+
+# See the get_class_ids function explanation
+auxiliary_id_map = {}
+
+
+class SnapchatMarketingException(Exception):
+    """ Just for formatting the exception as SnapchatMarketing"""
+
+
+class SnapchatMarketingStream(HttpStream, ABC):
+    data_field = None
+    url_base = "https://adsapi.snapchat.com/v1/"
+    primary_key = "id"
+
+    def __init__(self, start_date, **kwargs):
+        super().__init__(**kwargs)
+        self.start_date = pendulum.parse(start_date).to_rfc3339_string()
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page_cursor = response.json().get("paging", False)
+        if next_page_cursor:
+            return {"cursor": dict(parse_qsl(urlparse(next_page_cursor["next_link"]).query))["cursor"]}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return next_page_token if next_page_token else {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        if isinstance(self.data_field, list) and len(self.data_field) == 2:
+            json_response = response.json().get(self.data_field[0])
+            for resp in json_response:
+                yield resp.get(self.data_field[1])
+        else:
+            error_text = f"Cannot parse response. Data field unrecognized: {self.data_field}"
+            self.logger.error(error_text)
+            raise SnapchatMarketingException(error_text)
+
+
+class IncrementalSnapchatMarketingStream(SnapchatMarketingStream, ABC):
+    cursor_field = "updated_at"
+    data_field = None
+    cls_object = None
+    slice_key_name = "ad_account_id"
+
+    last_slice = None
+    current_slice = None
+    first_run = True
+    max_state = None
+
+    def stream_slices(self, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
+        stream_slices = self.get_class_ids(self.cls_object, self.slice_key_name, **kwargs)
+        if stream_slices:
+            self.last_slice = stream_slices[-1]
+
+        yield from stream_slices
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
How I check you filter records on client-side in `read_recoeds` method. So, you filter records after reading all records and didn't need control slicing with the state. For during first sync you pass all data from the start date and after you copy state to initial state and after reading (during the second sync, you will filter records by initial state).

```
    def read_records(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
        initial_state = copy.deepcopy(stream_state) or {}

        for record in super().read_records(stream_state=stream_state, **kwargs):
            if record.get(self.cursor_field, 0) >= initial_state.get(self.cursor_field, 0):
                yield record
```

I think that solution can simplify code in `get_updated_state`.

Reply:
Discussed on the meeting.